src/support/small_set.h

```diff
@@ -0,0 +1,280 @@
+    // movement in switching between fixed and flexible mode.
+    return flexible.empty();
+  }
+
+public:
+  using value_type = T;
+  using key_type = T;
+  using reference = T&;
+  using const_reference = const T&;
+  using set_type = FlexibleSet;
+  using size_type = size_t;
+
+  SmallSetBase() {}
+  SmallSetBase(std::initializer_list<T> init) {
+    for (T item : init) {
+      insert(item);
+    }
+  }
+
+  void insert(const T& x) {
+    if (usingFixed()) {
+      if (count(x)) {
+        // The item already exists, so there is nothing to do.
+        return;
+      }
+      // We must add a new item.
+      if (usedFixed < N) {
+        // Room remains in the fixed storage.
+        fixed[usedFixed++] = x;
+      } else {
+        // No fixed storage remains. Switch to flexible.
+        assert(usedFixed == N);
+        assert(flexible.empty());
+        flexible.insert(fixed.begin(), fixed.begin() + usedFixed);
+        flexible.insert(x);
+        assert(!usingFixed());
+        usedFixed = 0;
+      }
+    } else {
+      flexible.insert(x);
+    }
+  }
+
+  void erase(const T& x) {
+    if (usingFixed()) {
+      for (size_t i = 0; i < usedFixed; i++) {
+        if (fixed[i] == x) {
+          // We found the item; erase it by moving the final item to replace it
+          // and truncating the size.
+          usedFixed--;
+          fixed[i] = fixed[usedFixed];
+          return;
+        }
+      }
+    } else {
+      flexible.erase(x);
+    }
+  }
+
+  size_t count(const T& x) {
+    if (usingFixed()) {
+      // Do a linear search.
+      for (size_t i = 0; i < usedFixed; i++) {
+        if (fixed[i] == x) {
+          return 1;
+        }
+      }
+      return 0;
+    } else {
+      return flexible.count(x);
+    }
+  }
+
+  size_t size() const {
+    if (usingFixed()) {
+      return usedFixed;
+    } else {
+      return flexible.size();
+    }
+  }
+
+  bool empty() const { return size() == 0; }
+
+  void clear() {
+    usedFixed = 0;
+    flexible.clear();
+  }
+
+  bool operator==(const SmallSetBase<T, N, FlexibleSet>& other) const {
+    if (usingFixed()) {
+      if (usedFixed != other.usedFixed) {
+        return false;
+      }
+      for (size_t i = 0; i < usedFixed; i++) {
+        bool found = false;
+        for (size_t j = 0; j < usedFixed; j++) {
+          if (fixed[i] == other.fixed[j]) {
+            found = true;
+            break;
+          }
+        }
+        if (!found) {
+          return false;
+        }
+      }
+      return true;
+    }
+    return flexible == other.flexible;
+  }
+
+  bool operator!=(const SmallSetBase<T, N, FlexibleSet>& other) const {
+    return !(*this == other);
+  }
+
+  // iteration
+
+  template<typename Parent, typename FlexibleIterator> struct IteratorBase {
+    using iterator_category = std::forward_iterator_tag;
+    using difference_type = long;
+    using value_type = const T;
+    using pointer = value_type*;
+    using reference = value_type&;
```

Comment:
```suggestion
    using value_type = T;
    using pointer = const value_type*;
    using reference = const value_type&;
```

Reply:
I don't know enough C++ here - what would use `value_type`, that wants to know the type without `const`? (Maybe an obvious question)

Reply:
The elements of the set should be immutable, so we should only ever be handing out `const` pointers or references to them. If some generic code needs to recover the non-const-qualified type, it can use `std::remove_const`.

Reply:
I'm confused by the response here, sorry... Do we not hand out `const` pointer and references either way?

About the diff, if there isn't a need for someone to get a non-const `value_type`, then it seems safer before the diff..? But if the diff is just "proper C++" then sgtm too...

Reply:
I don't think this will matter for the execution here given that we don't provide any method that can modify things, but I think it is a recommended way: https://en.c
![image](https://user-images.githubusercontent.com/8726997/135325341-0ba091ec-0501-4e84-8596-4459528f957f.png)
ppreference.com/w/cpp/iterator/iterator_traits ("Specializations" section)

Also `value_type` is meant to be a value type without cv-qualifiers. It looks C++20 explicitly removes them.

Reply:
@kripken, disregard my previous comment; I think I was confused about what you were asking. @aheejin nice, I didn't know that before.

Reply:
Thanks @aheejin , interesting... sgtm, applying the diff.